SparseGrids/gridtestExternalTests.cpp

```diff
@@ -54,6 +54,30 @@ std::vector<double> genRandom(int num_samples, int num_dimensions = 1){
     for(auto &v : x) v = unif(park_miller);
     return x;
 }
+double unitDerivativeTests(const BaseFunction *f, TasmanianSparseGrid &grid) {
+    // Test the differentiate method on some common 1D nodes (in the 1st dimension) and return the max relative error.
+    std::vector<double> unique_nodes = {0.0, -1.0, 1.0, -0.5, 0.5, -0.5773502691896257, 0.5773502691896257};
+    int num_dimensions = f->getNumInputs();
+    int num_outputs = f->getNumOutputs();
+    std::vector<double> points(Utils::size_mult(num_dimensions, unique_nodes.size()), 0.0);
+    for (size_t i=0; i<unique_nodes.size(); i++) {
+        points[Utils::size_mult(i, num_dimensions)] = unique_nodes[i];
+    }
+    std::vector<double> err(unique_nodes.size(), 0.0), nrm(unique_nodes.size(), 0.0);
+    std::vector<double> r(num_outputs * num_dimensions), y(num_outputs * num_dimensions);
+    for (size_t k=0; k<unique_nodes.size(); k++) {
+        f->getDerivative(&(points[Utils::size_mult(k, num_dimensions)]), r.data());
+        grid.differentiate(&(points[Utils::size_mult(k, num_dimensions)]), y.data());
+        for (int i=0; i<num_outputs*num_dimensions; i++) {
+            nrm[k] = std::max(nrm[k], std::fabs(r[i]));
```

Comment:
Instead of storing `err` and `nrm` into vectors, define them as local constants in the for-loop for each k. Then you can merge the second for-loop into this one:
```
double nrm = 0.0;
double err = 0.0;
for(int i=0; i<num_outputs*num_dimensions; i++) { ... }
rel_err = std::max(rel_err, std::fabs(nrm) < Maths::num_tol ? err : err / nrm);
```

Reply:
Changed.